freeipa/src/main/java/com/sequenceiq/freeipa/service/image/ImageService.java

```diff
@@ -53,46 +65,111 @@ public Image create(Stack stack, ImageSettingsRequest imageRequest) {
+                () -> {
+                    LOGGER.debug("Default image not found. Attempt to find an image, compatible with the application version.");
+                    return filterImages(versions, compatibleImages, FreeIpaVersions::getImageIds);
+                }
+        ).or(
+                () -> {
+                    LOGGER.debug("Not found any image compatible with the application version. Falling back to the most recent image.");
+                    return compatibleImages.stream().max(Comparator.comparing(Image::getDate));
+                }
+        );
+    }
+
+    private Optional<Image> filterImages(List<FreeIpaVersions> freeIpaVersions, List<Image> images,
+            Function<FreeIpaVersions, List<String>> memberFunction) {
+        List<String> imageIds = freeIpaVersions.stream().map(memberFunction).flatMap(Collection::stream).collect(Collectors.toList());
+        return images.stream().filter(image -> imageIds.contains(image.getUuid())).max(Comparator.comparing(Image::getDate));
+    }
+
+    private List<FreeIpaVersions> filterFreeIpaVersionsByAppVersion(List<FreeIpaVersions> freeIpaVersions) {
+        List<FreeIpaVersions> exactFreeIpaVersionsMatches = freeIpaVersions.stream().filter(toExactVersionMatch()).collect(Collectors.toList());
+        if (!exactFreeIpaVersionsMatches.isEmpty()) {
+            LOGGER.debug("Exact version match found in image catalog for app version: {}", freeIpaVersion);
+            return exactFreeIpaVersionsMatches;
+        }
+        List<FreeIpaVersions> prefixFreeIpaVersions = freeIpaVersions.stream().filter(toPrefixVersionMatch()).collect(Collectors.toList());
+        if (!prefixFreeIpaVersions.isEmpty()) {
+            LOGGER.debug("Prefix version match found in image catalog for app version: {}", freeIpaVersion);
+            return prefixFreeIpaVersions;
         }
-        return image.get();
+
+        LOGGER.warn("Not found matching version in image catalog. Falling back to most recent image.");
+        return freeIpaVersions;
+    }
+
+    private Predicate<? super FreeIpaVersions> toPrefixVersionMatch() {
+        return freeIpaVersions -> freeIpaVersions.getVersions().stream().anyMatch(
+                version -> {
+                    Matcher appVersionMatcher = VERSION_PATTERN.matcher(freeIpaVersion);
+                    if (!appVersionMatcher.matches() || appVersionMatcher.groupCount() != 2) {
+                        return false;
+                    }
+                    String appVersionrefix = appVersionMatcher.group(1);
+                    Matcher versionPrefixMatcher = VERSION_PATTERN.matcher(version);
+                    if (!versionPrefixMatcher.matches() || versionPrefixMatcher.groupCount() != 2) {
+                        return false;
+                    }
+                    String versionPrefix = versionPrefixMatcher.group(1);
+                    return appVersionrefix.equals(versionPrefix);
+                });
```

Comment:
it's a bit hard to understand at first glance what's happening here, so I would suggest some refactor to make it easier.
so `appVersionrefix` (which is misspelled) could come from a method and could be optional. similarly `versionPrefix` could come from a well named method and finally you could have a return like `appVersionrefix.isPresent() && versionPrefix.isPresent() && appVersionrefix.get().equals(versionPrefix.get())`